### PR TITLE
liquidationV2 & auctionV2 draft

### PR DESCRIPTION
**Pull request by pratikasr**

### CODE REVIEW

File: proto/comdex/liquidationsV2/v1beta1/liquidate.proto [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    (gogoproto.moretags) = "yaml:\"is_english_activated\""];
+
+  EnglishAuctionParam english_auction_param = 10[
+    (gogoproto.customname) = "EnglishAuctionParam",
+    (gogoproto.moretags) = "yaml:\"english_auction_param\""];
+  //One thing to keep in mind that somehow it should not happen that a void is created where something at level 2 gets triggerred and it has no data saved a level 1 for lookup and it fails .
+
+
+}
+
```
Comment (dheerajkd30) [inline proto/comdex/liquidationsV2/v1beta1/liquidate.proto:65]: Auction params are specific to apps only or have we thought asset specific parameters also ?

Comment (cybertronprime) [reply proto/comdex/liquidationsV2/v1beta1/liquidate.proto:65]: There exist two types of auction params, one is for app-level that can be controlled by them, and the other one is comdex level which protocol will control.
Bid.proto in auctionsV2 proto contains - protocol level param - AuctionParams
liquidate.proto in liquidationsV2  proto contains - app level param - LiquidationWhiteListing

**Comment by dheerajkd30** [comment]
Resolve the conflicts please.

File: proto/comdex/auctionsV2/v1beta1/auction.proto [hunk 2]
```diff
@@ -0,0 +1,112 @@
+syntax = "proto3";
+package comdex.auctionsV2.v1beta1;
```
Comment (dheerajkd30) [inline proto/comdex/auctionsV2/v1beta1/auction.proto:2]: the better and accepted directory structure would be comdex.auctions.v2. Let changes this across @pratikasr @rajrohit10

**Comment by cybertronprime** [comment]
Working on it

File: proto/comdex/auctionsV2/v1beta1/auction.proto (hunk 2 above)
Comment (pratikasr) [reply proto/comdex/auctionsV2/v1beta1/auction.proto:2]: Updated

File: x/auctionsV2/keeper/bid.go [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+
+		if !totalCollateralTokenQuanitity.LTE(auctionData.CollateralToken.Amount) {
+			//This means that there is less collateral available .
+			leftOverCollateral := auctionData.CollateralToken.Amount
+			_, debtTokenAgainstLeftOverCollateral, _ := k.vault.GetAmountOfOtherToken(ctx, auctionData.CollateralAssetId, auctionData.CollateralTokenAuctionPrice, leftOverCollateral.Sub(collateralTokenQuanitityForBonus), auctionData.DebtAssetId, debtPrice)
+			bid.Amount = debtTokenAgainstLeftOverCollateral
+			totalCollateralTokenQuanitity = leftOverCollateral
+			//Amount to call from reserve account for adjusting the auction target debt
+			//So we call the module account to give funds to compensate the user.
+			debtGettingLeft := auctionData.DebtToken.Sub(sdk.NewCoin(auctionData.DebtToken.Denom, debtTokenAgainstLeftOverCollateral))
```
Comment (dheerajkd30) [inline x/auctionsV2/keeper/bid.go:59]: debtGettingLeft  -> leftOverDebt

File: x/auctionsV2/keeper/bid.go [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+	}
+	//Check to update bid.Amount
+	fullBid := false
+
+	if bid.Amount.GTE(auctionData.DebtToken.Amount) {
+		bid.Amount = auctionData.DebtToken.Amount
+		fullBid = true
+
+	}
+	_, collateralTokenQuanitity, _ := k.vault.GetAmountOfOtherToken(ctx, auctionData.DebtAssetId, debtPrice, bid.Amount, auctionData.CollateralAssetId, auctionData.CollateralTokenAuctionPrice)
```
Comment (dheerajkd30) [inline x/auctionsV2/keeper/bid.go:43]: GetAmountOfOtherToken ??

**Comment by cybertronprime** [comment]
Using this from previous auctio module to calculate the token quantity of
other token from the given token. Nomen taken from auction v1

File: x/auctionsV2/keeper/bid.go [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+		//Creating user bid struct
+		biddingId, err := k.CreateUserBid(ctx, auctionData.AppId, bidder, auctionID, sdk.NewCoin(auctionData.CollateralToken.Denom, totalCollateralTokenQuanitity), sdk.NewCoin(auctionData.DebtToken.Denom, bid.Amount), "dutch")
+		if err != nil {
+			return bidId, err
+		}
+		//Based on app type call perform specific function - external , internal and /or keeper incentive
+		//See if this was keeper initiated transaction- then incentivisation will be in place based on the percentage
+		//For apps that are external to comdex chain
+
+		if liquidationData.InitiatorType == "external" {
```
Comment (dheerajkd30) [inline x/auctionsV2/keeper/bid.go:118]: use switch case here please.

File: x/liquidationsV2/keeper/liquidate.go [hunk 6]
```diff
@@ -0,0 +1,732 @@
...
+
+		//Checking if the vault getting liquidated is a cmst based vault or not
+		//This is primarily to infer that primary market will consider cmst at $1 at the time of buying it
+		isCMST := !extPair.AssetOutOraclePrice
+
+		//Creating locked vault struct , which will trigger auction
+		//This function will only trigger Dutch auction
+		//before creating locked vault, checking that Dutch auction is already there in the whitelisted liquidation data
+		if !whitelistingData.IsDutchActivated {
+			return fmt.Errorf("Error , dutch auction not activated by the app, this function is only to trigger dutch auctions %d", whitelistingData.IsDutchActivated)
```
Comment (dheerajkd30) [inline x/liquidationsV2/keeper/liquidate.go]: error format is wrong type.

**Comment by cybertronprime** [comment]
Already corrected it in the new pr across both modules. Will push in some
time.

**Review by dheerajkd30 — CHANGES_REQUESTED** [review]
resolve the comments , need to refactor few functions, will update on team call.

**Review by dheerajkd30 — APPROVED** [review]

File: app/app.go [hunk 7]
```diff
@@ -767,6 +786,54 @@ func New(
 		&app.LendKeeper,
 	)
 
+	app.NewliqKeeper = newliqkeeper.NewKeeper(
```
Comment (dheerajkd30) [inline app/app.go:788]: duplicates ? line 821

File: x/auctionsV2/abci.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/comdex-official/comdex/x/auctionsV2/keeper"
+)
+
+func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
+	err := k.AuctionIterator(ctx)
+	if err != nil {
+		return
```
Comment (dheerajkd30) [inline x/auctionsV2/abci.go]: error logs and event emmiter .

File: x/auctionsV2/abci.go [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+	}
+	err = k.LimitOrderBid(ctx)
+	if err != nil {
+		return
```
Comment (dheerajkd30) [inline x/auctionsV2/abci.go]: error logs and event emmiter

**Review by cgsingh33 — APPROVED** [review]

**Review by dheerajkd30 — APPROVED** [review]

**Pull request merged**